packages/plugins/typescript/typed-document-sdk/src/sdk-base.ts

```diff
@@ -0,0 +1,383 @@
+  DocumentNode,
+  FieldNode,
+  Kind,
+  ListTypeNode,
+  NamedTypeNode,
+  NonNullTypeNode,
+  OperationTypeNode,
+  SelectionNode,
+  SelectionSetNode,
+  TypeNode,
+  VariableDefinitionNode,
+} from 'graphql';
+// IMPORTS END
+
+type Mutable<T> = { -readonly [P in keyof T]: T[P] };
+
+type Impossible<K extends keyof any> = {
+  [P in K]: never;
+};
+
+/** Do not allow any other properties as the ones defined in the base type. */
+type NoExtraProperties<T, U extends T = T> = U & Impossible<Exclude<keyof U, keyof T>>;
+
+/** Require at least one property in the object defined. */
+type AtLeastOnePropertyOf<T> = {
+  [K in keyof T]: { [L in K]-?: T[L] } & { [L in Exclude<keyof T, K>]?: T[L] };
+}[keyof T];
+
+/**
+ * @source https://stackoverflow.com/a/56874389/4202031
+ */
+type KeysMatching<T extends { [key: string]: any }, V> = {
+  [K in keyof T]-?: T[K] extends V ? K : never;
+}[keyof T];
+
+type SDKSelection = { [key: string]: any };
+
+type ResultType = { [key: string]: any };
+
+type SDKOperationType<TSelection extends SDKSelection, TType extends ResultType> = {
+  [TKey in keyof TSelection]: TKey extends keyof TType
+    ? TType[TKey] extends Record<string, any>
+      ? SDKOperationType<TSelection[TKey], TType[TKey]>
+      : TType[TKey]
+    : never;
+};
+
+export type SDKSelectionSet<TType extends Record<string, any> = any> = AtLeastOnePropertyOf<NoExtraProperties<TType>>;
+
+type SDKInputTypeMap = { [inputTypeName: string]: any };
+
+type SDKArgumentType<
+  T_SDKInputTypeMap extends SDKInputTypeMap,
+  T_VariableDefinitions extends SDKVariableDefinitions<T_SDKInputTypeMap>
+> = { [T_VariableName in keyof T_VariableDefinitions]: T_SDKInputTypeMap[T_VariableDefinitions[T_VariableName]] };
+
+type SDKSelectionTypedDocumentNode<
+  T_Selection extends SDKSelection,
+  T_ResultType extends ResultType,
+  T_SDKInputTypeMap extends SDKInputTypeMap | void,
+  T_VariableDefinitions extends SDKVariableDefinitions<
+    T_SDKInputTypeMap extends void ? never : T_SDKInputTypeMap
+  > | void
+> = TypedDocumentNode<
+  SDKOperationType<T_Selection, T_ResultType>,
+  T_SDKInputTypeMap extends SDKInputTypeMap
+    ? T_VariableDefinitions extends SDKVariableDefinitions<T_SDKInputTypeMap>
+      ? SDKArgumentType<T_SDKInputTypeMap, T_VariableDefinitions>
+      : never
+    : never
+>;
+
+export const SDKFieldArgumentSymbol = Symbol('SDKFieldArguments');
+
+type SDKInputNonNullType<T extends string> = `${T}!`;
+type SDKInputListType<T extends string> = `[${T}]`;
+
+/**
+ * Poor mans implementation, this should actually be recursive as you could potentially indefinitely nest non nullable and list types...
+ * Right now we only allow Type, [Type], [Type]!, [Type!] and [Type!]!
+ */
+type SDKInputContainerType<T extends string> =
+  | T
+  | SDKInputNonNullType<T>
+  | SDKInputListType<T>
+  | SDKInputNonNullType<SDKInputListType<T>>
+  | SDKInputListType<SDKInputNonNullType<T>>
+  | SDKInputNonNullType<SDKInputListType<SDKInputNonNullType<T>>>;
```

Comment:
TODO: Investigate whether this can be done more flexible with https://devblogs.microsoft.com/typescript/announcing-typescript-4-5/#tailrec-conditional